File: apps/api-rs/src/document/pdf.rs
use super::plain_text::split_paragraphs;
use super::shared::finalize_parsed;
use super::*;
pub(super) fn parse_pdf(
    doc_id: Uuid,
    parse_job_id: Uuid,
    title: &str,
    bytes: &[u8],
) -> Result<ParsedDocument> {
    let page_count = pdf_page_sizes(bytes)
        .map(|sizes| sizes.len())
        .unwrap_or_default();
    if page_count > MAX_PDF_PAGES {
        bail!("pdf_page_count_exceeded:{}>{}", page_count, MAX_PDF_PAGES);
    }

    let pages = pdf_extract::extract_text_from_mem_by_pages(bytes)
        .or_else(|_| {
            pdf_extract::extract_text_from_mem(bytes).map(|text| {
                if text.contains('\x0C') {
                    text.split('\x0C').map(str::to_string).collect()
                } else {
                    vec![text]
                }
            })
        })
        .context("pdf_text_extract_failed")?;

    if pages.len() > MAX_PDF_PAGES {
        bail!("pdf_page_count_exceeded:{}>{}", pages.len(), MAX_PDF_PAGES);
    }

    let mut blocks = Vec::new();
    let mut anchors = Vec::new();
    let mut warnings = Vec::new();
    let mut current_heading: Option<String> = None;
    let mut pages_without_text = 0usize;

    for (page_idx, page_text) in pages.iter().enumerate() {
        let page = (page_idx + 1) as i32;
        let page_chars = page_text.chars().count();
        if page_chars > MAX_PDF_PAGE_TEXT_CHARS {
            bail!("pdf_page_text_chars_exceeded:{page}:{page_chars}>{MAX_PDF_PAGE_TEXT_CHARS}");
        }
        let paragraphs = split_paragraphs(page_text);
        if paragraphs.is_empty() {
            pages_without_text += 1;
            warnings.push(format!("pdf_page_{}_no_text_layer", page));
        }

        for paragraph in &paragraphs {
            let trimmed = paragraph.trim();
            if trimmed.is_empty() {
                continue;
            }
            let is_heading = looks_like_heading(trimmed);
            let block_id = Uuid::new_v4();
            let anchor = SourceAnchor::for_pdf_paragraph(
                doc_id,
                parse_job_id,
                Uuid::nil(), // tenant_id 将在上层填充
                block_id,
                page,
                trimmed,
                None,
            );
            let anchor_id = anchor.anchor_id;
            anchors.push(anchor);

            let heading_path = current_heading.iter().cloned().collect::<Vec<_>>();
            blocks.push(ParsedBlock {
                block_id,
                block_index: blocks.len() as i32,
                block_type: if is_heading { "heading" } else { "paragraph" }.to_string(),
                text: trimmed.to_string(),
                heading_level: is_heading.then_some(1),
                heading_path: if is_heading { vec![] } else { heading_path },
                page_start: Some(page),
                page_end: Some(page),
                slide_index: None,
                table_id: None,
                bbox: None,
                anchor_ids: vec![anchor_id],
                source_ref: json!({"format": "pdf", "page": page}),
                metadata: json!({"layout": "text_layer"}),
            });
            if is_heading {
                current_heading = Some(trimmed.to_string());
            }
        }
    }
    if blocks.is_empty() {
        warnings.push("scanned_pdf_no_text_layer".to_string());
    } else if pages_without_text > 0 {
        warnings.push(format!(
            "pdf_partial_text_layer:{}/{}",
            pages_without_text,
            pages.len()
        ));
    }
    let mut parsed = finalize_parsed(
        doc_id,
        parse_job_id,
        "pdf",
        title,
        Some(pages.len() as i32),
        blocks,
        vec![],
        anchors,
    );
    parsed.warnings.extend(warnings);
    Ok(parsed)
}

fn looks_like_heading(text: &str) -> bool {
    let char_count = text.chars().count();
    if !(2..=80).contains(&char_count)
        || text.ends_with(['.', '。', '！', '!', '？', '?', '；', ';', '，', ','])
    {
        return false;
    }
    let meaningful = text.chars().filter(|c| c.is_alphanumeric()).count();
    let digits = text.chars().filter(|c| c.is_ascii_digit()).count();
    meaningful > 0 && digits.saturating_mul(2) < meaningful
}

/// 使用 lopdf 提取每页的 CropBox/MediaBox 尺寸（点）。
fn pdf_page_sizes(bytes: &[u8]) -> Result<Vec<(f64, f64)>> {
    let doc = lopdf::Document::load_mem(bytes).context("lopdf_load_failed")?;
    let pages = doc.get_pages();
    let mut sizes = Vec::new();
    for page_num in 1..=pages.len() as u32 {
        let page_id = pages
            .get(&page_num)
            .copied()
            .ok_or_else(|| anyhow!("lopdf_missing_page_{}", page_num))?;
        let bounds = page_bounds(&doc, page_id);
        if let lopdf::Object::Array(arr) = bounds {
            if arr.len() >= 4 {
                let x0 = as_f64(&arr[0]).unwrap_or(0.0);
                let y0 = as_f64(&arr[1]).unwrap_or(0.0);
                let x1 = as_f64(&arr[2]).unwrap_or(595.0);
                let y1 = as_f64(&arr[3]).unwrap_or(842.0);
                sizes.push(((x1 - x0).abs(), (y1 - y0).abs()));
            }
        }
    }
    Ok(sizes)
}

fn page_bounds(doc: &lopdf::Document, page_id: lopdf::ObjectId) -> lopdf::Object {
    let default = lopdf::Object::Array(vec![
        lopdf::Object::Integer(0),
        lopdf::Object::Integer(0),
        lopdf::Object::Integer(595),
        lopdf::Object::Integer(842),
    ]);
    let Ok(page) = doc.get_dictionary(page_id) else {
        return default;
    };
    let resolve = |obj: &lopdf::Object| doc.get_object(obj.as_reference().unwrap_or((0, 0))).ok();
    page.get(b"CropBox")
        .ok()
        .and_then(resolve)
        .or_else(|| page.get(b"MediaBox").ok().and_then(resolve))
        .cloned()
        .unwrap_or(default)
}

fn as_f64(obj: &lopdf::Object) -> Option<f64> {
    match obj {
        lopdf::Object::Integer(i) => Some(*i as f64),
        lopdf::Object::Real(f) => Some((*f).into()),
        _ => None,
    }
}

File: apps/api-bun/src/document/pdf.ts
import { getDocument } from 'pdfjs-dist/legacy/build/pdf.mjs';
import { WorkerMessageHandler } from 'pdfjs-dist/legacy/build/pdf.worker.mjs';

import { newUuid } from '../infra/uuid.ts';
import {
  sourceAnchorForPdfParagraph,
  sourceAnchorTableCellRange,
  type SourceAnchor,
} from '../models/source_anchor.ts';
import { finalizeParsed } from './shared.ts';
import { charCount, trimRust } from './text_utils.ts';
import {
  NIL_UUID,
  MAX_PDF_PAGES,
  MAX_PDF_PAGE_TEXT_CHARS,
  type ParsedBlock,
  type ParsedDocument,
  type ParsedTable,
  type ParsedTableCell,
} from './types.ts';
import { extractPdfPageLayout, type PdfLayoutBlock, type PdfPageLayout } from './pdf_layout.ts';

const pdfjsGlobal = globalThis as typeof globalThis & {
  pdfjsWorker?: { WorkerMessageHandler: typeof WorkerMessageHandler };
};
pdfjsGlobal.pdfjsWorker ??= { WorkerMessageHandler };

export async function parsePdf(
  docId: string,
  parseJobId: string,
  title: string,
  bytes: Uint8Array,
): Promise<ParsedDocument> {
  const { task, doc } = await loadDocument(bytes);
  try {
    if (doc.numPages > MAX_PDF_PAGES) {
      throw new Error(`pdf_page_count_exceeded:${doc.numPages}>${MAX_PDF_PAGES}`);
    }
    const pages = await extractPageLayouts(doc);
    const blocks: ParsedBlock[] = [];
    const tables: ParsedTable[] = [];
    const anchors: SourceAnchor[] = [];
    const warnings: string[] = [];
    const repeatedNoise = repeatedHeaderFooterText(pages);
    let currentHeading: string | null = null;
    let pagesWithoutText = 0;

    for (const [pageIndex, layout] of pages.entries()) {
      const page = pageIndex + 1;
      const pageChars = charCount(layout.text);
      if (pageChars > MAX_PDF_PAGE_TEXT_CHARS) {
        throw new Error(`pdf_page_text_chars_exceeded:${page}:${pageChars}>${MAX_PDF_PAGE_TEXT_CHARS}`);
      }
      if (layout.blocks.length === 0) {
        pagesWithoutText += 1;
        warnings.push(`pdf_page_${page}_no_text_layer`);
        continue;
      }
      const bodyFonts = layout.blocks
        .filter((block) => block.kind === 'text' && block.text.length >= 20)
        .map((block) => block.font_size);
      const bodyFont = median(bodyFonts.length > 0 ? bodyFonts : layout.blocks.map((block) => block.font_size));

      for (const layoutBlock of layout.blocks) {
        const text = trimRust(layoutBlock.text);
        if (!text || repeatedNoise.has(noiseKey(layoutBlock))) continue;
        if (layoutBlock.kind === 'table' && layoutBlock.rows) {
          const result = createPdfTable(docId, parseJobId, page, layoutBlock, tables.length, blocks.length, currentHeading);
          tables.push(result.table);
          blocks.push(result.block);
          anchors.push(result.anchor);
          continue;
        }

        const blockId = newUuid();
        const formula = layoutBlock.kind === 'formula';
        const heading = !formula && looksLikeHeading(text, layoutBlock.font_size, bodyFont);
        const blockType = formula ? 'formula' : heading ? 'heading' : 'paragraph';
        const anchor = sourceAnchorForPdfParagraph(
          docId, parseJobId, NIL_UUID, blockId, page, text, layoutBlock.bbox,
        );
        anchor.kind = formula ? 'formula' : heading ? 'heading' : 'paragraph';
        anchor.source_ref = {
          format: 'pdf', page, pdf_text_run_ids: layoutBlock.source_run_ids,
        };
        anchors.push(anchor);
        blocks.push({
          block_id: blockId,
          block_index: blocks.length,
          block_type: blockType,
          text,
          heading_level: heading ? 1 : null,
          heading_path: heading ? [] : currentHeading ? [currentHeading] : [],
          page_start: page,
          page_end: page,
          slide_index: null,
          table_id: null,
          bbox: layoutBlock.bbox,
          anchor_ids: [anchor.anchor_id],
          source_ref: anchor.source_ref,
          metadata: {
            layout: 'positioned_text_layer',
            font_size: layoutBlock.font_size,
            extraction_method: 'text_layer',
            ...(formula ? { formula_format: 'plain_text' } : {}),
          },
        });
        if (heading) currentHeading = text;
      }
    }

    if (blocks.length === 0) warnings.push('scanned_pdf_no_text_layer');
    else if (pagesWithoutText > 0) warnings.push(`pdf_partial_text_layer:${pagesWithoutText}/${pages.length}`);
    if (tables.length > 0) warnings.push(`pdf_tables_detected:${tables.length}`);

    const parsed = finalizeParsed(docId, parseJobId, 'pdf', title, pages.length, blocks, tables, anchors);
    parsed.warnings.push(...warnings);
    return parsed;
  } finally {
    try {
      await task.destroy();
    } catch (error) {
      console.error('[documind][document] pdf worker cleanup failed:', messageOf(error));
    }
  }
}

async function loadDocument(bytes: Uint8Array) {
  const task = getDocument({ data: bytes.slice(), useSystemFonts: false });
  try {
    return { task, doc: await task.promise };
  } catch (error) {
    throw new Error('pdf_text_extract_failed:' + messageOf(error));
  }
}

async function extractPageLayouts(doc: {
  numPages: number;
  getPage(pageNumber: number): Promise<{
    getViewport(options: { scale: number }): { width: number; height: number };
    getTextContent(): Promise<{ items: unknown[] }>;
  }>;
}): Promise<PdfPageLayout[]> {
  try {
    const pages: PdfPageLayout[] = [];
    for (let pageNumber = 1; pageNumber <= doc.numPages; pageNumber += 1) {
      pages.push(await extractPdfPageLayout(await doc.getPage(pageNumber), pageNumber));
    }
    return pages;
  } catch (error) {
    throw new Error('pdf_text_extract_failed:' + messageOf(error));
  }
}

function createPdfTable(
  docId: string,
  parseJobId: string,
  page: number,
  layout: PdfLayoutBlock,
  tableIndex: number,
  blockIndex: number,
  currentHeading: string | null,
): { table: ParsedTable; block: ParsedBlock; anchor: SourceAnchor } {
  const rows = layout.rows ?? [];
  const width = Math.max(0, ...rows.map((row) => row.length));
  const normalizedRows = rows.map((row) => Array.from({ length: width }, (_, index) => row[index] ?? ''));
  const blockId = newUuid();
  const tableId = newUuid();
  const cells: ParsedTableCell[] = [];
  for (const [rowIndex, row] of normalizedRows.entries()) {
    for (const [columnIndex, text] of row.entries()) {
      cells.push({
        cell_id: newUuid(), row_index: rowIndex, col_index: columnIndex,
        rowspan: 1, colspan: 1, text, normalized_text: text.replace(/\s+/g, ' ').trim(),
        is_header: rowIndex === 0, data_type: tableDataType(text), bbox: null, style: {},
        source_ref: { format: 'pdf', page, table_index: tableIndex, row: rowIndex, col: columnIndex },
      });
    }
  }
  const headers = normalizedRows[0] ?? [];
  const body = normalizedRows.slice(1);
  const cellRange = {
    row_start: 0,
    row_end: Math.max(0, normalizedRows.length - 1),
    col_start: 0,
    col_end: Math.max(0, width - 1),
  };
  const sourceRef = {
    format: 'pdf', page, table_index: tableIndex, table_id: tableId,
    pdf_text_run_ids: layout.source_run_ids, cell_range: cellRange,
  };
  const anchor = sourceAnchorTableCellRange(
    docId, parseJobId, NIL_UUID, 'pdf', blockId, tableId, page, null,
    cellRange, sourceRef, layout.text, layout.bbox,
  );
  const headingPath = currentHeading ? [currentHeading] : [];
  return {
    table: {
      table_id: tableId, block_id: blockId, table_index: tableIndex,
      title: currentHeading, heading_path: headingPath,
      page_start: page, page_end: page, slide_index: null,
      headers, rows: body, cells, markdown: layout.text,
      quality: { header_confidence: 0.7, grid_confidence: layout.confidence ?? 0.65, extraction_method: 'text_alignment' },
      source_ref: sourceRef,
    },
    block: {
      block_id: blockId, block_index: blockIndex, block_type: 'table', text: layout.text,
      heading_level: null, heading_path: headingPath,
      page_start: page, page_end: page, slide_index: null, table_id: tableId,
      bbox: layout.bbox, anchor_ids: [anchor.anchor_id], source_ref: sourceRef,
      metadata: { layout: 'positioned_table', extraction_method: 'text_layer', table_grid_confidence: layout.confidence ?? 0.65 },
    },
    anchor,
  };
}

function repeatedHeaderFooterText(pages: PdfPageLayout[]): Set<string> {
  const pagesByKey = new Map<string, Set<number>>();
  for (const [pageIndex, page] of pages.entries()) {
    for (const block of page.blocks) {
      if (block.kind !== 'text' || (block.bbox.y1 < 0.88 && block.bbox.y0 > 0.12)) continue;
      const key = noiseKey(block);
      if (key.length < 2 || key.length > 160) continue;
      const seen = pagesByKey.get(key) ?? new Set<number>();
      seen.add(pageIndex);
      pagesByKey.set(key, seen);
    }
  }
  const minimum = Math.max(2, Math.ceil(pages.length * 0.6));
  return new Set([...pagesByKey].filter(([, pageSet]) => pageSet.size >= minimum).map(([key]) => key));
}

function noiseKey(block: PdfLayoutBlock): string {
  return block.text.replace(/\d+/g, '#').replace(/\s+/g, ' ').trim().toLowerCase();
}

function looksLikeHeading(text: string, fontSize: number, bodyFont: number): boolean {
  const length = charCount(text);
  if (length < 2 || length > 100) return false;
  const last = [...text].pop();
  if (last !== undefined && '.。！!？?；;，,'.includes(last)) return false;
  let meaningful = 0;
  let digits = 0;
  for (const ch of text) {
    if (/[\p{L}\p{N}]/u.test(ch)) meaningful += 1;
    if (ch >= '0' && ch <= '9') digits += 1;
  }
  const numbered = /^(第[一二三四五六七八九十百]+[章节]|\d+(?:\.\d+)*[、.\s])/u.test(text);
  return meaningful > 0 && digits * 2 < meaningful && (numbered || fontSize >= bodyFont * 1.18);
}

function tableDataType(text: string): string {
  const value = text.trim();
  if (/^[+-]?(?:\d{1,3}(?:,\d{3})*|\d+)(?:\.\d+)?%$/u.test(value)) return 'percentage';
  if (/^[¥￥$€£]?\s*[+-]?(?:\d{1,3}(?:,\d{3})*|\d+)(?:\.\d+)?$/u.test(value)) return 'number';
  if (/^\d{4}[-/.年]\d{1,2}(?:[-/.月]\d{1,2}日?)?$/u.test(value)) return 'date';
  return 'text';
}

function median(values: number[]): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[middle]! : (sorted[middle - 1]! + sorted[middle]!) / 2;
}

function messageOf(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
